keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -3,111 +3,556 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
 ## Summary
 
-The Kubernetes API server has always read service account keys from disk as the process starts, and kept them in memory for the duration of the server's lifetime. As the API server can now verify and issue projected volume tokens, it would be advantageous to support external signing and verifying of token data over an API, as well as reading public keys from an API.
+Service account keys are used by kube-apiserver for JWT signing and authentication. Keys are stored on the disk and are loaded by kube-apiserver during process start time. Key management for service account keys is entirely within the cluster and is less flexible in that the signing keys remain the same for the lifetime of the process. This KEP proposes updates that will allow Kubernetes distributions to integrate with key management solutions of their choice (eg: HSMs, cloud KMSes).  
 
 ## Motivation
 
-For operators who want to regularly rotate the signing and verifying keys for projected volume tokens, the Kubernetes API server must be restarted in order to use a new key. To facilitate easy key rotation, this KEP includes an proposal for a grpc API to support out of process signing and listing of signing keys.
+1. Ease of rotation.
+  At present, kube-apiserver loads service account keys at process start time and the keys remain the same for the life-time of the process. Thus, rotating keys require a kube-apiserver to restart. External key management eliminates the need to restart.
+
+2. Enhanced security.
+  kube-apiserver loads service account keys from a file on the disk. This means anyone with privileged access to the kube-apiserver configuration / filesystem could exfiltrate signing material. An external signer that never returns signing material would mitigate this risk.
 
 ### Goals
 
-- Support for out-of-process JWT signing
-- Support for listing public verifying keys
+- Enable integration with HSMs and cloud KMSes
+- Support out-of-process JWT signing
+- Support listing public verifying keys
+- Ensure tokens signed by external signers remain consistent with tokens issued by kube-apiserver (consistent headers, algorithms, claims, etc)
 - Preserve existing behavior and performance for keys not read over a socket
+<<[UNRESOLVED]>>
+- Move legacy token signing from kube-controller-manager to kube-apiserver and thus external signing.
```

Comment:
Still chewing on what to do about this bit. The interaction between the controller that produces non-expiring tokens and an external signer that can rotate out signing keys or limit the max token lifetime is pretty weird...

Folks using external signers have a couple reasonable options I see:
1. Turn off the legacy token controller loop and choose not to support non-expiring legacy tokens in their cluster
2. a) Populate legacy token secrets with externally signed tokens *and* b) augment token controller logic with some awareness of when the tokens need to be replaced in those secrets.

1 is easy and can be accomplished by not giving kube-controller-manager a signing key.

2a _could_ be done by constructing the existing token controller loop with an alternate TokenGenerator implementation that talks to an external signer, 2b requires new bookkeeping and rotation logic.

For 2, I'm on the fence about whether it is worth it to try to teach kube-controller-manager about external signers and token rotation, or if most external signers will turn off legacy tokens or have their own rotation logic that would make a generic kube-controller-manager implementation less useful to them.

Reply:
> I think we should move the existing controller to a staging repo, have it take an interface for the signer that can be implemented with any external signer, and leave it at that. That gives three choices to users:
>
> 1. Run the built-in signer with a file based signing key that is never rotated
> 2. Don't run the built-in signer or any custom re-implementation
> 3. Don't run the built-in signer and instead run a custom re-implementation using the staging repo

For option 1, would that mean their external signer impl would have to stitch in that public key for verification in its returned keyset?

2 seems ideal for clusters which can handle dropping legacy tokens capability

3 seems ~ok to me for clusters which cannot handle dropping legacy tokens capability

Reply:
> I think we should move the existing controller to a staging repo, have it take an interface for the signer that can be implemented with any external signer, and leave it at that.

I'd be on board with that.

Reply:
For alpha, settled on not doing anything in-tree for the legacy token controller, and documenting the options for folks using external signers as:
1. Turn off legacy token support
2. Keep legacy token controller in `kube-controller-manager` running using a file-based signing key and have the external signer return the legacy public key (optionally with `exclude_from_oidc_discovery: true`)
3. Implement and run a custom external controller to sign legacy tokens using an external signer

Add a beta graduation criteria to get feedback from folks who would choose option 3 on whether they would reuse legacy token controller code if it were located to a staging repo and allowed to take a token signer interface. If feedback indicates this would be valuable to the community, relocate that code as part of the beta graduation criteria.

Reply:
called out the Legacy token scenario is a section ("Support for Legacy Tokens") under "Notes/Constraints/Caveats" section